### PR TITLE
use aten parallel on lu factor

### PR DESCRIPTION
**Pull request by zhuhaozhe**
https://github.com/pytorch/pytorch/issues/91536. One issue mentioned torch.inv is pretty slow for large batches with small matrices on cuda.

I checked the CPU implementations and found we have an optimize opportunity.
For torch.inv, the CPU pass chooses to solve it by `lu_factor` + `lu_solve`.
The `lu_factor` loop on `batch_size` dimension and the parallel happened inside lapack
 - For small matrix, the computational complexity is too tiny to parallel inside lapack.
 - Even for large matrix, the parallelization efficiency is not good in lapack ( it performs worse than using at::parallel outside) 
 - Only for small batch size + small matrix size, the omp overhead will take too large overhead.

Based on the above observations, using at::parallel outside on lu_factor will have a pretty large benefit.

Here is the code/data collected on 32 core ICX system.
```python
import torch
import time

def bench(bs, r):
    x = torch.randn(int(bs), r, r)
    start = time.time()
    for i in range(100):
        y1 = torch.linalg.lu_factor(x)
    end = time.time()
    print(r, bs)
    print(end - start)
    print((end - start)/(r**3))

for r in (4, 16, 64):
    for bs in (1e2, 1e4, 1e6):
        bench(bs, r)
```

| bs/rank | 100/4 |  10000/4 |  1000000/4 | 100/16 |  10000/16|  1000000/16| 100/64|  10000/64|  1000000/64|
| ---- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
| parallel inside lapack | 0.0028 |1.077 | 11.99|0.0163 | 1.5260|153.17 |0.2021|20.93 | 1877|
| parallel outside lapack | 0.0087 | 0.0247 | 1.566| 0.0044|0.1678 |17.63|0.038|2.311 | 208.6|
|speed up ratio| 0.32x | 43.6x  | 7.65x|3.70x |9.09x |8.69x |5.32x |9.06x |9x |

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
How does performance change with `MKL_DYNAMIC=TRUE/FALSE` and varying `MKL_NUM_THREADS`?

**Comment by IvanYashchuk** [comment]
How does performance compare to `getrf_batch_strided` https://www.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-fortran/top/lapack-routines/lapack-linear-equation-routines/lapack-linear-equation-computational-routines/matrix-factorization-lapack-computational-routines/getrf-batch-strided.html?

**Comment by zhuhaozhe** [comment]
> How does performance change with `MKL_DYNAMIC=TRUE/FALSE` and varying `MKL_NUM_THREADS`?

There is no significant difference for setting `MKL_DYNAMIC=FALSE`.

Varying `MKL_NUM_THREADS` (bs = 1000000 are removed since it is too slow and bs=10000 is enough to see the gap):
For `thread_num = 1`, the perf are same.
For `thread_num = 4`:
| bs/rank | 100/4 |  10000/4 | 100/16 |  10000/16|100/64|  10000/64|
| ---- | --- | --- | --- | --- | --- | --- |
| parallel inside lapack | 0.0025 |0.1040 |0.0208|1.768 | 0.182|23.55|
| parallel outside lapack | 0.0038 | 0.0489 | 0.0082| 0.625|0.0667 |8.226|
|speed up ratio| 0.65x | 2.13x|2.54x |2.83x |2.73x |2.86x |

For `thread_num = 16`:
| bs/rank | 100/4 |  10000/4 | 100/16 |  10000/16|100/64|  10000/64|
| ---- | --- | --- | --- | --- | --- | --- |
| parallel inside lapack | 0.0025 |0.1023 | 0.0207|1.63 | 0.1814|21.02|
| parallel outside lapack | 0.0063 | 0.0287 | 0.0058| 0.2155|0.0232 |3.234|
|speed up ratio| 0.39x | 3.56x|3.56x |7.56x |7.82x |6.50x |

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp [hunk 1]
```diff
@@ -915,11 +914,30 @@ void apply_lu_factor(const Tensor& input, const Tensor& pivots, const Tensor& in
...
+  int64_t matrix_size = std::min(m, n);
+  int64_t computational_complexity =
+      batch_size * matrix_size * matrix_size * matrix_size;
+  // A threshold test out on 32 core/socket ICX system
+  bool do_parallel = computational_complexity >= 102400;
+  if (do_parallel) {
+    at::parallel_for(0, batch_size, 0, loop);
+  } else {
+    loop(0, batch_size);
   }
```
Comment (nikitaved) [inline aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:941]: Hmm, shouldn't we swap the order? Seems like the heavier the matrix, the better it is for LAPACK do things in parallel with serial batch loop, or am I missing something?

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp [hunk 2]
```diff
@@ -915,11 +914,30 @@ void apply_lu_factor(const Tensor& input, const Tensor& pivots, const Tensor& in
...
+          n,
+          input_working_ptr,
+          leading_dimension,
+          pivots_working_ptr,
+          infos_working_ptr);
+    }
+  };
+  int64_t matrix_size = std::min(m, n);
+  int64_t computational_complexity =
+      batch_size * matrix_size * matrix_size * matrix_size;
```
Comment (nikitaved) [inline aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:934]: What if this overflows?

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp [hunk 3]
```diff
@@ -915,11 +914,30 @@ void apply_lu_factor(const Tensor& input, const Tensor& pivots, const Tensor& in
   auto n = input.size(-1);
   auto leading_dimension = std::max<int64_t>(1, m);
 
-  for (const auto i : c10::irange(batch_size)) {
-    scalar_t* input_working_ptr = &input_data[i * input_matrix_stride];
-    int* pivots_working_ptr = &pivots_data[i * pivots_stride];
-    int* infos_working_ptr = &infos_data[i];
-    lapackLu<scalar_t>(m, n, input_working_ptr, leading_dimension, pivots_working_ptr, infos_working_ptr);
+  std::function<void(int64_t, int64_t)> loop = [](int64_t, int64_t) {};
+  loop = [&](int64_t start, int64_t end) {
```
Comment (nikitaved) [inline aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:918]: nit: `const auto loop = [&](...) { ... }` but up to you.

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp (hunk 1 above)
Comment (jgong5) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:941]: Can we do parallel on the batch dim and devise a good grain size per batch to let `at::parallel_for` judge whether to do multi-threading or serial execution?

Comment (zhuhaozhe) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:941]: Hi, @nikitaved and @jgong5.
The data shows while the matrix gets heavier(rank=64), parallel on small batch size (100) will still get a large speed up (>5x). So whether to use parallel is not only related to batch size but also related to matrix size. So I use the `computational_complexity` to decide whether use at::parallel outside. Maybe `computational_complexity size/current thread num` can reflect the speed up more precisely, but since there is almost a speed up on all thread num cases so we may not need to divide by thread num.

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp (hunk 2 above)
Comment (zhuhaozhe) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:934]: Thanks for pointing this out, will use log-space.
Sizes will not be zero, there is a short circuit.
```
  if (A.numel() == 0) {
    // zero out the infos as it will have one element if the input is a matrix of size (0, 0)
    info.zero_();
    return;
  }
```

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp (hunk 3 above)
Comment (zhuhaozhe) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:918]: Will change

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp (hunk 1 above)
Comment (nikitaved) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp]: What about matrices of sizes >= 1024, for example? I speculate that doing all-serial batch loop is better in this case, and cache is expected to be utilized more efficiently...

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp [hunk 4]
```diff
@@ -915,11 +914,29 @@ void apply_lu_factor(const Tensor& input, const Tensor& pivots, const Tensor& in
...
+          n,
+          input_working_ptr,
+          leading_dimension,
+          pivots_working_ptr,
+          infos_working_ptr);
+    }
+  };
+  int64_t matrix_size = std::min(m, n);
+  int64_t computational_complexity =
+      std::log(batch_size) + 3 * std::log(matrix_size);
```
Comment (nikitaved) [inline aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:933]: `std::log` returns a floating point.

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp (hunk 1 above)
Comment (jgong5) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp]: How about devise the grain size according to the matrix size? The grain size is smaller when the matrix size is bigger. And we always do at::parallel_for on the batch dim.

> I speculate that doing all-serial batch loop is better in this case, and cache is expected to be utilized more efficiently... To generalize, it seems like there is a matrix size threshold past which it is always better to delegate parallelization to the linalg routine.

According to the PR description, it seems worse to do parallelization inside lapack: "Even for large matrix, the parallelization efficiency is not good in lapack ( it performs worse than using at::parallel outside)". Not sure why it is the case.

Comment (nikitaved) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp]: The largest matrix in the benchmark is just 64 by 64.

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp (hunk 3 above)
Comment (lezcano) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp]: Yep, using `std::function` explicitly creates unnecessary overhead over just using a lambda via `auto`

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp (hunk 1 above)
Comment (zhuhaozhe) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp]: Hi, @nikitaved. I tried matrix size up to 32k * 32k and parallel in aten won't worse than parallel inside lapck.
| bs/rank | 32/1024 |  32/4096 | 32/8192 |  32/16384 |32/32768|
| ---- | --- | --- | --- | --- | --- | 
|speed up ratio| 2.75x | 1.23x|1.02x |1.02x |1.03x |

Comment (zhuhaozhe) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp]: > How about devise the grain size according to the matrix size? The grain size is smaller when the matrix size is bigger. And we always do at::parallel_for on the batch dim.
> 
> > I speculate that doing all-serial batch loop is better in this case, and cache is expected to be utilized more efficiently... To generalize, it seems like there is a matrix size threshold past which it is always better to delegate parallelization to the linalg routine.
> 
> According to the PR description, it seems worse to do parallelization inside lapack: "Even for large matrix, the parallelization efficiency is not good in lapack ( it performs worse than using at::parallel outside)". Not sure why it is the case.

Yes, now we are using matrix size to decide grain_size.

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp [hunk 5]
```diff
@@ -915,12 +914,29 @@ void apply_lu_factor(const Tensor& input, const Tensor& pivots, const Tensor& in
...
+    }
+  };
+  int64_t matrix_rank = std::min(m, n);
+  // A threshold test out on 32 core/socket ICX system
+  int64_t grain_size;
+  float log_grain_size = std::log(3200) - 3 * std::log(matrix_rank);
+  if (log_grain_size < 0)
+    grain_size = 1;
+  else
+    grain_size = int64_t(std::exp(log_grain_size));
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:938]: Taking logs and then exps is not only slow but it may also lose precision.
```suggestion
  auto matrix_rank = std::min(m, n);
  // A heuristic tested on a 32 core/socket ICX system
  // https://github.com/pytorch/pytorch/pull/93037#discussion_r1090112948
  auto grain_size = int64_t{3200.0 / (matrix_rank * matrix_rank * matrix_rank)}
```

also, more generally, please use braces when using if-else clauses.

**Review by lezcano — APPROVED** [review]
Just a minor comment. In general it's good to leave a link to the benchmarks for future reference.

Otherwise this LGTM.

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp (hunk 5 above)
Comment (nikitaved) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:938]: We used log to avoid overflows, and now the result is being exped to again introduce potential overflow issues...

Comment (nikitaved) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:938]: I read it like this: the larger the matrix, the more weight it is to parallelize over batch dimensions. Is this really what we want?

Comment (lezcano) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:938]: It should be, as per https://github.com/pytorch/pytorch/pull/93037#issuecomment-1405007779, right?

Comment (nikitaved) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:938]: As I have mentioned above, these benchmarks are just for matrices up to 64, I would prefer to see a more comprehensive test case for different dtypes and sizes.

Comment (zhuhaozhe) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp]: Hi, @nikitaved. I updated the benchmark for bs=32, thread_num=32 (since we set the grain size >=1). And iter the rank from 1024 to 16384.

| rank | 1k |  2k |  4k | 8k |  16k| 
| ---- | --- | --- | --- | --- | --- |
| parallel inside lapack | 1.0476 |2.7907| 11.163|52.867 |358.62 |
| parallel outside lapack | 0.4165 | 1.6828 | 9.1045| 52.794| 347.25 |
|speed up ratio| 2.52x | 1.61x  | 1.22xx|1.001x |1.03x |

It is true "the larger the matrix, the more weight it is for parallelizing over batch dimensions inside lapack". But since we see parallel within `at::parallel` will not worse than parallel inside lapack. So we may not need to set an upper bound to back to parallel inside lapack.
 



```python
# lu_factor.py
import torch
import time

def bench(bs, r):
    x = torch.randn(int(bs), r, r)
    start = time.time()
    for i in range(10):
        y1 = torch.linalg.lu_factor(x)
    end = time.time()
    print(end - start)

for r in [10, 11, 12, 13, 14]:
    rr = 2**r
    for bs in [32]:
        bench(bs, rr)
```
```bash
export LD_PRELOAD=${CONDA_PREFIX}/lib/libjemalloc.so
echo "parallel inside"
OMP_NUM_THREADS=32 numactl -C 0-31 -m 0 python -u lu_factor.py
echo "parallel outside"
LU_TORCH_PARALLEL=1 OMP_NUM_THREADS=32 numactl -C 0-31 -m 0 python -u lu_factor.py
```

Comment (zhuhaozhe) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp]: > Taking logs and then exps is not only slow but it may also lose precision (not in this computation, but still).
> 
> also, more generally, please use braces when using if-else clauses.

Thanks, done as your suggestions. But still remain the grain_size should >= 1.

File: aten/src/ATen/native/BatchLinearAlgebraKernel.cpp (hunk 1 above)
Comment (nikitaved) [reply aten/src/ATen/native/BatchLinearAlgebraKernel.cpp]: @zhuhaozhe , the benchmarks are just for `float`, what about `double` and complex types?
